File: src/model/nostr/network/subscriptions/cache_processor.go
//go:build js && wasm
// +build js,wasm

package subscriptions

import (
	"context"
	"fmt"

	"github.com/candypoets/nutscash/nostr/types"
	"github.com/nbd-wtf/go-nostr"
	"github.com/rs/zerolog"
)

const (
	MaxRecursionDepth = 2
)

// cacheProcessor implements the CacheProcessor interface
type cacheProcessor struct {
	database EventDatabase
	parser   EventParser
	logger   zerolog.Logger
}

// NewCacheProcessor creates a new cache processor
func NewCacheProcessor(database EventDatabase, parser EventParser, logger zerolog.Logger) CacheProcessor {
	return &cacheProcessor{
		database: database,
		parser:   parser,
		logger:   logger,
	}
}

// ProcessLocalRequests searches for events in the database based on requests,
// processes them, and recursively handles any resulting requests up to a maximum depth.
func (cp *cacheProcessor) ProcessLocalRequests(
	ctx context.Context,
	requests []types.Request,
	depth int,
) ([]types.Request, [][]types.ParsedEvent, error) {
	if depth >= MaxRecursionDepth {
		cp.logger.Debug().
			Int("depth", depth).
			Msg("Reached maximum recursion depth for local requests")
		return make([]types.Request, 0), make([][]types.ParsedEvent, 0), nil
	}

	filteredRequests, events, err := cp.database.QueryEventsForRequests(requests, depth > 0)
	if err != nil {
		cp.logger.Warn().
			Err(err).
			Int("depth", depth).
			Msg("Error querying database")
		return filteredRequests, nil, fmt.Errorf("database query failed: %w", err)
	}

	processedEvents := make([][]types.ParsedEvent, 0, len(events))

	for _, event := range events {
		select {
		case <-ctx.Done():
			return filteredRequests, processedEvents, ctx.Err()
		default:
		}

		// Store relay hint if present
		cp.parser.GetRelayHint(event.Event)

		// Parse event if not already parsed
		if event.Parsed == nil {
			parsedEvent, err := cp.parser.Parse(nostr.Event{
				ID:        event.ID,
				Kind:      event.Kind,
				CreatedAt: event.CreatedAt,
				Tags:      event.Tags,
				Content:   event.Content,
				PubKey:    event.PubKey,
				Sig:       event.Sig,
			})
			if err != nil {
				cp.logger.Warn().
					Err(err).
					Str("event_id", event.ID).
					Msg("Error parsing event from database")
				continue
			}
			event = parsedEvent
		}

		eventsWithContext := []types.ParsedEvent{event}

		// Handle recursive requests from parsed event
		if event.Requests != nil && len(*event.Requests) > 0 {
			contextEvents := cp.FindEventContext(event, 3)
			// Append child events to the current event's context
			eventsWithContext = append(eventsWithContext, contextEvents...)
		}

		processedEvents = append(processedEvents, eventsWithContext)
	}

	return filteredRequests, processedEvents, nil
}

// FindEventContext builds context for an event by recursively finding related events
func (cp *cacheProcessor) FindEventContext(event types.ParsedEvent, maxDepth int) []types.ParsedEvent {
	var context []types.ParsedEvent
	cp.findEventContextRecursive(event, &context, 0, maxDepth, make(map[string]bool))
	return context
}

// findEventContextRecursive is the internal recursive function for building event context
func (cp *cacheProcessor) findEventContextRecursive(
	event types.ParsedEvent,
	context *[]types.ParsedEvent,
	depth int,
	maxDepth int,
	visited map[string]bool,
) {
	if depth > maxDepth {
		return
	}

	// Avoid infinite loops by tracking visited events
	if visited[event.ID] {
		return
	}
	visited[event.ID] = true

	// Process requests from this event
	if event.Requests != nil && len(*event.Requests) > 0 {
		for _, request := range *event.Requests {
			relatedEvents, err := cp.database.QueryEvents(request.ToFilter())
			if err != nil {
				cp.logger.Warn().
					Err(err).
					Str("event_id", event.ID).
					Msg("Error querying related events")
				continue
			}

			for _, relatedEvent := range relatedEvents {
				*context = append(*context, relatedEvent)
				cp.findEventContextRecursive(relatedEvent, context, depth+1, maxDepth, visited)
			}
		}
	}
}

// GetEventsByFilter is a helper method to query events by filter
func (cp *cacheProcessor) GetEventsByFilter(filter nostr.Filter) ([]types.ParsedEvent, error) {
	return cp.database.QueryEvents(filter)
}

// ValidateEvent performs basic validation on an event
func (cp *cacheProcessor) ValidateEvent(event types.ParsedEvent) error {
	if event.ID == "" {
		return fmt.Errorf("event ID cannot be empty")
	}
	if event.PubKey == "" {
		return fmt.Errorf("event public key cannot be empty")
	}
	if event.CreatedAt == 0 {
		return fmt.Errorf("event created_at cannot be zero")
	}
	return nil
}

// ShouldCacheEvent determines if an event should be cached based on its kind
func (cp *cacheProcessor) ShouldCacheEvent(event types.ParsedEvent) bool {
	cachableKinds := map[int]bool{
		0:     true, // Metadata events
		3:     true, // Contact lists
		4:     true, // Direct messages
		10002: true, // Relay list metadata
		10019: true, // nuts.cash user settings
		17375: true, // nuts.cash encrypted wallet event
	}
	return cachableKinds[event.Kind]
}

// GetStatistics returns processing statistics
func (cp *cacheProcessor) GetStatistics() CacheProcessorStats {
	// This would be implemented with actual metrics collection
	return CacheProcessorStats{
		EventsProcessed:  0,
		CacheHits:        0,
		CacheMisses:      0,
		RecursiveQueries: 0,
	}
}

// CacheProcessorStats represents processing statistics
type CacheProcessorStats struct {
	EventsProcessed  int64
	CacheHits        int64
	CacheMisses      int64
	RecursiveQueries int64
}

File: src/model/nostr-worker/src/network/subscriptions/cache_processor.rs
use crate::db::NostrDB;
use crate::network::subscriptions::interfaces::{
    CacheProcessor as CacheProcessorTrait, EventDatabase,
};
use crate::parser::Parser;
use crate::types::network::Request;
use crate::types::*;
use anyhow::Result;
use async_trait::async_trait;
use std::collections::HashSet;
use std::sync::Arc;
use tracing::{debug, warn};

pub struct CacheProcessor {
    database: Arc<NostrDB>,
    parser: Arc<Parser>,
}

impl CacheProcessor {
    pub fn new(database: Arc<NostrDB>, parser: Arc<Parser>) -> Self {
        Self { database, parser }
    }

    pub async fn find_context_events_simple(
        &self,
        event: &ParsedEvent,
        max_depth: usize,
    ) -> Vec<ParsedEvent> {
        let mut context_events = Vec::new();
        let mut visited = HashSet::<String>::new();

        self.find_event_context_recursive(event, &mut context_events, 0, max_depth, &mut visited)
            .await;

        context_events
    }

    async fn find_event_context_recursive(
        &self,
        event: &ParsedEvent,
        context: &mut Vec<ParsedEvent>,
        depth: usize,
        max_depth: usize,
        visited: &mut HashSet<String>,
    ) {
        if depth > max_depth {
            return;
        }

        // Avoid infinite loops by tracking visited events
        if visited.contains(&event.event.id.to_hex()) {
            return;
        }
        visited.insert(event.event.id.to_hex());

        // Process requests from this event (matching Go implementation)
        if let Some(requests) = &event.requests {
            for request in requests {
                if let Ok(filter) = request.to_filter() {
                    if let Ok(related_events) =
                        EventDatabase::query_events(&*self.database, filter).await
                    {
                        for related_event in related_events {
                            context.push(related_event.clone());
                            Box::pin(self.find_event_context_recursive(
                                &related_event,
                                context,
                                depth + 1,
                                max_depth,
                                visited,
                            ))
                            .await;
                        }
                    }
                }
            }
        }
    }

    fn create_event_filter(&self, event_id: &str) -> Result<nostr::Filter> {
        let event_id = nostr::EventId::from_hex(event_id)?;
        Ok(nostr::Filter::new().id(event_id))
    }

    fn create_profile_filter(&self, pubkey: &str) -> Result<nostr::Filter> {
        let pubkey = nostr::PublicKey::from_hex(pubkey)?;
        Ok(nostr::Filter::new()
            .author(pubkey)
            .kind(nostr::Kind::Metadata))
    }

    fn create_address_filter(&self, address: &str) -> Result<nostr::Filter> {
        // Parse address format: kind:pubkey:d_tag
        let parts: Vec<&str> = address.split(':').collect();
        if parts.len() != 3 {
            return Err(anyhow::anyhow!("Invalid address format"));
        }

        let kind = parts[0].parse::<u64>()?;
        let pubkey = nostr::PublicKey::from_hex(parts[1])?;
        let d_tag = parts[2];

        Ok(nostr::Filter::new()
            .kind(nostr::Kind::from(kind))
            .author(pubkey)
            .custom_tag(
                nostr::SingleLetterTag::lowercase(nostr::Alphabet::D),
                vec![d_tag],
            ))
    }
}

#[async_trait(?Send)]
impl CacheProcessorTrait for CacheProcessor {
    async fn process_local_requests(
        &self,
        requests: Vec<Request>,
        max_depth: usize,
    ) -> Result<(Vec<Request>, Vec<Vec<ParsedEvent>>)> {
        debug!(
            "Processing {} local requests with max depth {}",
            requests.len(),
            max_depth
        );

        let mut remaining_requests = Vec::new();
        let mut cached_events_batches = Vec::new();

        for request in requests {
            // Convert request to filter and query local database
            match request.to_filter() {
                Ok(filter) => {
                    match EventDatabase::query_events(&*self.database, filter).await {
                        Ok(events) => {
                            if events.is_empty() {
                                debug!("No cached events found for request");
                                // No cached events, add to remaining requests
                                remaining_requests.push(request);
                                // cached_events_batches.push(Vec::new());
                            } else {
                                debug!("Found {} cached events for request", events.len());
                                // Found cached events
                                // Process each event and build context like Go implementation
                                let mut processed_events: Vec<Vec<ParsedEvent>> = Vec::new();

                                for event in &events {
                                    // Store the main event
                                    let mut events_with_context = vec![event.clone()];

                                    // Handle recursive requests from parsed event (like Go)
                                    if !request.no_context
                                        && event.requests.is_some()
                                        && !event.requests.as_ref().unwrap().is_empty()
                                    {
                                        let context_events =
                                            self.find_context_events_simple(event, 3).await;
                                        events_with_context.extend(context_events);
                                    }

                                    processed_events.push(events_with_context);
                                }

                                cached_events_batches.extend(processed_events);

                                // Check if we need to fetch more from network
                                if !request.cache_first {
                                    remaining_requests.push(request);
                                }
                            }
                        }
                        Err(e) => {
                            warn!("Error querying local events: {}", e);
                            remaining_requests.push(request);
                            cached_events_batches.push(Vec::new());
                        }
                    }
                }
                Err(e) => {
                    warn!("Error converting request to filter: {}", e);
                    remaining_requests.push(request);
                    cached_events_batches.push(Vec::new());
                }
            }
        }

        debug!(
            "Found {} cached event batches, {} remaining requests",
            cached_events_batches.len(),
            remaining_requests.len()
        );

        Ok((remaining_requests, cached_events_batches))
    }

    async fn find_event_context(&self, event: &ParsedEvent, _max_depth: usize) -> Vec<ParsedEvent> {
        // Simplified implementation to avoid Send trait issues
        let mut context_events = Vec::new();

        // Process requests from this event (matching Go implementation)
        if let Some(requests) = &event.requests {
            for request in requests {
                if let Ok(filter) = request.to_filter() {
                    if let Ok(related_events) =
                        EventDatabase::query_events(&*self.database, filter).await
                    {
                        context_events.extend(related_events);
                    }
                }
            }
        }

        context_events
    }
}
